src/go/k8s/controllers/redpanda/cluster_controller.go

```diff
@@ -89,8 +91,18 @@ func (r *ClusterReconciler) Reconcile(
 
 	var observedPods corev1.PodList
 
-	err := r.List(ctx, &observedPods, &client.ListOptions{
-		LabelSelector:	labels.ForCluster(&redpandaCluster).AsClientSelector(),
+	newReq, err := labels2.NewRequirement(
```

Comment:
I am sorry, I am still unsure that we should be handling this for the user :(

Kubernetes is built around notion of label selectors which is set of labels that identifies set of objects. You typically use this to group your pods that belong to the same service etc. For us that set of resource is name + instance + component. If someone creates another instance with SAME NAME under SAME NAMESPACE, that's a user error. The same way you would be able to create another pod that belongs to the service.

I know there is no way how to enforce unique-ness of instance name in namespace, kubernetes only ensures unique-ness of resource name. Perhaps the fix then is to ALWAYS replace instance name with resource name and then it's resolved, it can never clash. Right now we replace it for owned objects only if it's not provided, but perhaps we should enforce that. When we have admission webhook we can make it even more obvious by rejecting objects that specify instance but the instance is not same as resource name.

WDYT?

Reply:
> For us that set of resource is name + instance + component.

For now I agree, but couldn't we change that?

> If someone creates another instance with SAME NAME under SAME NAMESPACE, that's a user error.

I don't think I fallow. You can not create resource with the same name in the same namespace. 

> The same way you would be able to create another pod that belongs to the service.

Agree that if someone creates pod with the same set of labels it will intervene with our reporting of nodes. 

 > Perhaps the fix then is to ALWAYS replace instance name with resource name and then it's resolved, it can never clash. Right now we replace it for owned objects only if it's not provided, but perhaps we should enforce that.

From my point of view UX is then broken. As a user I would expect to define labels in custom resource and other objects should follow my intention. Then I can relay on retrieving resources from k8s API server.

> When we have admission webhook we can make it even more obvious by rejecting objects that specify instance but the instance is not same as resource name.

That is the argument that convince me to not push for this PR.

I would add only that this change doesn't introduce more complexity. It moves more responsibility to the operator which IMHO is good.

@dimitriscruz I would say you need to be arbiter for our dispute.
@alenkacz Should we leave the golang e2e tests?

Reply:
Let's not give up just yet, sorry if my comment felt like I am hard blocking this PR from merge. I think we're still discussing. I thought about it a bit more I actually don't have a huge problem with the solution, it just opened a bigger topic that we could perhaps talk about.

I find it problematic that someone can so easily override our instance label and I think all kinds of stuff break then (thanks for pointing it out! I did not see how huge problem it is but now I do). We use that selector as selector in multiple objects, I think we should make sure they are unique or other stuff might break as well.

Two solutions come to my mind that would also solve the pods case in this PR:
- force override the instance label to name of the resource and add validation webhook that fails object creation if that is not true
- add one more label to our selector, something like `vectorized.io/redpanda-instance: this-will-always-be-resouce-name`

Thoughts? :)

Reply:
I must be missing something. Under `labels.go` we do `labels[InstanceKey] = cluster.Name`. Cluster name corresponds to the resource name that the user provides, right? Given a name, there can only be one resource of each kind in each namespace with that name. How are we ending up with this issue again?

Reply:
@dimitriscruz 
> Under labels.go we do labels[InstanceKey] = cluster.Name. Cluster name corresponds to the resource name that the user provides, right? Given a name, there can only be one resource of each kind in each namespace with that name. How are we ending up with this issue again?

It's correct, but after you create default set of labels the [`merge` function](https://github.com/vectorizedio/redpanda/blob/1f9316723013b345a2ad9e080026d3ac9b2c77a0/src/go/k8s/pkg/labels/labels.go#L73-L89) will overwrite this if user set that label.

@alenkacz 
> force override the instance label to name of the resource and add validation webhook that fails object creation if that is not true

To rephrase what you write: Let's create validation webhook that fails at the object creation if it tries overwrite our 3 key labels.

> add one more label to our selector, something like `vectorized.io/redpanda-instance: this-will-always-be-resouce-name`

This doesn't looks much different from what I propose. We are managing `statefulset` right now, which solves tracing it's own resources by putting `controller-revision-hash` label to track its own pods and pods during update.

For reference [this change introduces update method for statefulset](https://github.com/kubernetes/kubernetes/commit/1b55f57391f5a648b1f9859118a2d1fef7017ffe) and [this comment describe how we should tract/report available nodes to the end user based on statfulset update strategy](https://github.com/kubernetes/kubernetes/blob/b2b8c1f18d1056db23c1cd377d6dc3dd4fb4dcc1/pkg/controller/statefulset/stateful_set_control.go#L262-L270)

Reply:
oh I see, neat, I did not know about that. That changes my perspective  and makes me wonder whether we won't need such a hash as well. I think your PR solves part of the problem (the pod selection for status reporting) but I think if people create this duplicate label I expect that other selectors might be causing problems as well (we use that for service, volumes etc.)

okay I am fine with that direction then, let's please link that STS reference to the code. I'll make another pass of review later today.

Thanks for the explanation and opening this topic!